api.js

```diff
@@ -834,18 +873,69 @@ const createTimetable = function(obj) {
     });
 };
 
-const getSuggestions = function(timetable) {
+const getUserAvailableTimes = function(timetable) {
+    return new Promise(function(resolve, reject) {
+        let times = [];
+        //iterate through timetable and find first 5 suggestions
+        for(var i = 0; i < timetable.length; i++) {
+            if(timetable[i].available == 0) {
+                times.push(timetable[i]);
+            }
+        }
+        resolve(times);
+    });
+};
+
+const createRoomSuggestions = function(userTimes, roomMeetings) {
+    console.log(userTimes);
+    console.log(roomMeetings);
+
+    return new Promise(function(resolve, reject) {
+        for(var x = 0; x < userTimes.length; x++) {
+            var timeSlot = userTimes[x];
+            for(var y = 0; y < roomMeetings.length; y++) {
+                var room = roomMeetings[y];
+                //console.log(room);
+                var roomName = room.room_name;
+                var meetingTimes = (room.meetingTimes).split("|");
+                timeSlot.rooms.push(roomName);
+                for(var i = 0; i < meetingTimes.length - 1; i+=2) {
+
+                    var startTime = meetingTimes[i];
+                    var endTime = meetingTimes[i+1];
+
+                    if((((new Date(timeSlot.startDateTime)).getTime() <= (new Date(startTime)).getTime()) && 
+                    ((new Date(timeSlot.endDateTime)).getTime() > (new Date(startTime)).getTime())) ||
+                   (((new Date(timeSlot.startDateTime)).getTime() <= (new Date(endTime)).getTime()) && 
+                    ((new Date(timeSlot.endDateTime)).getTime() >= (new Date(endTime)).getTime()))) {
```

Comment:
Not sure if these inequalities are correct

Reply:
there is a lot of duplicate calls of Date().getTime() that we should be able to reduce.

IIRC the pseudocode should look like
```
if((meetingStart >= timeSlotStart && meetingStart <= timeSlotEnd) 
|| (meetingEnd > timeSlotStart && meetingEnd <= timeSlotEnd) 
|| (meetingStart <= timeSlotStart && meetingEnd >= timeSlotEnd)) {
// slot not available
} else {
  // slot available for this meeting
}
```

Reply:
another way to refactor would be to put this conditional into a function that returns a boolean.  then we can evaluate the conditional on its own with test data.  This is a really good way to unit test complex conditionals; separating the conditional from the logic that is utilizing it.